src/modules/flight_mode_manager/tasks/AutoPrecisionLanding/FlightTaskAutoPrecisionLanding.cpp

```diff
@@ -0,0 +1,472 @@
+
+	_constraints.want_takeoff = _checkTakeoff();
+
+	return ret;
+}
+
+void
+FlightTaskAutoPrecisionLanding::run_state_start()
+{
+	// In this state simply track the navigator setpoint triplet
+	// This ensures that the behaviour for precision landing during RTL / LAND
+	// is the same as without precision landing.
+	// This flight task will generate independent setpoints once the vehicle
+	// is in the vertical landing phase.
+	_position_setpoint = _target;
+
+	if (switch_to_state_horizontal_approach()) {
+		// If target visible and go to horizontal approach directly
+		return;
+
+	} else if ((PrecLandMode)_param_rtl_pld_md.get() == PrecLandMode::Opportunistic) {
+		// could not see the target immediately, so just fall back to normal landing
+		switch_to_state_fallback();
+
+	} else if (_type == WaypointType::land) {
+		// Navigator already entered land stage. Take over with precision landing
+		// This is where the vehicle starts to behave differently than in regular land!
+		switch_to_state_search();
+	}
+}
+
+void
+FlightTaskAutoPrecisionLanding::run_state_horizontal_approach()
+{
+	float x = _target_pose.x_abs;
+	float y = _target_pose.y_abs;
+	slewrate(x, y);
+
+	// Fly to target XY position, but keep navigator's altitude setpoint
+	_position_setpoint(0) = x;
+	_position_setpoint(1) = y;
+	_position_setpoint(2) = _target(2);
+
+	_velocity_setpoint(0) = _velocity_setpoint(1)  = NAN;
+
+	// If altitude setpoint is NAN, we stay at the current altitude
+	if (!PX4_ISFINITE(_position_setpoint(2))) {
+		_velocity_setpoint(2) = 0.0f;
+	}
+
+	// check if target visible, if not go to start
+	if (!check_state_conditions(PrecLandState::HorizontalApproach)) {
+		PX4_WARN("%s, state: %i", LOST_TARGET_ERROR_MESSAGE, (int) _state);
+
+		// Stay at current position for searching for the landing target
+		// TODO: This is not going to work after conversion to flight task, because
+		// the position_setpoint will be overwritten at the next iteration!
+		// Solution: An additional wait state
+		_position_setpoint = _position;
+		_velocity_setpoint(0) = _velocity_setpoint(1) = _velocity_setpoint(2) = NAN;
+
+		if (!switch_to_state_start()) {
+			switch_to_state_fallback();
+		}
+
+		return;
+	}
+
+	if (check_state_conditions(PrecLandState::DescendAboveTarget)) {
+		if (!_point_reached_time) {
+			_point_reached_time = hrt_absolute_time();
+		}
+
+		if (hrt_absolute_time() - _point_reached_time > 2000000) {
+			// if close enough for descent above target go to descend above target
+			if (switch_to_state_descend_above_target()) {
+
+				return;
+			}
+		}
+
+	}
+}
+
+void
+FlightTaskAutoPrecisionLanding::run_state_descend_above_target()
+{
+	// Overwrite Auto setpoints in order to descend above target
+	_position_setpoint(0) = _target_pose.x_abs;
+	_position_setpoint(1) = _target_pose.y_abs;
+	_position_setpoint(2) = NAN;
+	_velocity_setpoint(0) = 0;
+	_velocity_setpoint(1) = 0;
+	_velocity_setpoint(2) = _param_mpc_land_speed.get();
```

Comment:
This should use the same logic here for the different land speeds and heights
https://github.com/PX4/PX4-Autopilot/blob/master/src/modules/flight_mode_manager/tasks/Auto/FlightTaskAuto.cpp#L236-L245